File: BinaryTree/2.Binary Tree Level Order Traversal II(LeetCode).cpp
// Link -https://leetcode.com/problems/binary-tree-level-order-traversal/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        if(root == NULL) return ans;
        vector<int> partialAns;
        queue<TreeNode*> q;
        q.push(root);
        q.push(NULL);
        while(q.size() > 0){
            TreeNode * front = q.front();
            q.pop();
            if(front != NULL){
                partialAns.push_back(front->val);//print
                if(front->left != NULL){
                    q.push(front->left);
                }
                if(front->right != NULL){
                    q.push(front->right);
                }
            }
            else{
                ans.push_back(partialAns);
                if(q.size() > 0){
                    q.push(NULL);    
                    partialAns.clear();
                }
            }
        }
       
        return ans;
    }
};

File: BinaryTree/2.Binary Tree Level Order Traversal II (Leetcode).java
// Link -https://leetcode.com/problems/binary-tree-level-order-traversal/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if(root == null) {
            return ans;
        }
        List<Integer> partialAns = new ArrayList<>();
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        q.add(null);
        Stack<List<Integer>> st = new Stack<>();
        while(q.size() > 0) {
            TreeNode front = q.remove();// remove
            if(front != null){
                partialAns.add(front.val); // print
                if(front.left != null) {
                    q.add(front.left);
                }
                if(front.right != null) {
                    q.add(front.right);
                }
            }else{
                st.push(partialAns);
                if(q.size() > 0){
                    q.add(null);
                    partialAns = new ArrayList<>();
                }
            }
        }
        while(st.size() > 0){
            List<Integer> top = st.pop();
            ans.add(top);
        }
        return ans;
    }
}